pkg/storer/internal/pinning/pinning.go

```diff
@@ -264,11 +274,19 @@ func (c *collectionPutter) Close(st internal.Storage, writer storage.Writer, roo
 	c.mtx.Lock()
 	defer c.mtx.Unlock()
 
-	// Save the root pin reference.
-	c.collection.Addr = root
-	err := writer.Put(c.collection)
+	collection := &pinCollectionItem{Addr: root}
+	has, err := st.IndexStore().Has(collection)
 	if err != nil {
-		return fmt.Errorf("pin store: failed updating collection: %w", err)
+		return fmt.Errorf("pin store: check previous root: %w", err)
+	}
+
+	if !has {
+		// Save the root pin reference.
+		c.collection.Addr = root
+		err := writer.Put(c.collection)
+		if err != nil {
+			return fmt.Errorf("pin store: failed updating collection: %w", err)
+		}
 	}
```

Comment:
This doesn't solve the problem of residual collectionChunks with their corresponding forever-remaining increments of the chunkstore RefCnt.   Consider:

1) Two or more simultaneous operations start pinning the exact same content.
2) They all do the individual Puts, and the root collection doesn't yet exist, so they all actually store their collectionChunks and invoke chunkstore.Put incrementing RefCnt on those chunks.
3) One of the pinners actually executes the Close causing the pinCollectionItem to exist.  That one "wins" and actually completes the pin.
4) When the other pinners finally invoke Close, they'll correctly not replace the original pinCollectionItem, but all of those incorrect RefCnt increments are still there causing those chunks to never be able to be removed from the chunkstore.

Reply:
Basically returning an error on line :276 will make `cleanupOnErrWriter` wrapper object to perform a cleanup (by calling the `Cleanup` method) whenever the error is detected
https://github.com/ethersphere/bee/blob/bf56d82e1aff5d5e9fde395044a410e9b8baf03f/pkg/api/api.go#L870-L874
notice the `onErr` pointing to the `.Cleanup` method
https://github.com/ethersphere/bee/blob/bf56d82e1aff5d5e9fde395044a410e9b8baf03f/pkg/api/bzz.go#L106-L110
which in turn will decrement the ref count(s) to the correct value. See the test below.

Reply:
Thank you for that description.   I should have followed the error path back out from Close() myself.  I'm setting up a testnet environment to verify this on a full OSM tileset load whenever it is merged.